File: src/git.ts
/**
 * Git context — captured at build time via withInariWatch plugin,
 * read at runtime from injected env vars.
 */

export interface GitContext {
  commit: string
  branch: string
  message: string
  timestamp: string
  dirty: boolean
}

/**
 * Read git context from env vars injected by withInariWatch at build time.
 * Returns null if no git context is available (e.g., no .git directory).
 */
export function getGitContext(): GitContext | null {
  if (typeof process === "undefined" || !process.env) return null
  const commit = process.env.INARIWATCH_GIT_COMMIT
  if (!commit) return null

  return {
    commit,
    branch: process.env.INARIWATCH_GIT_BRANCH || "unknown",
    message: process.env.INARIWATCH_GIT_MESSAGE || "",
    timestamp: process.env.INARIWATCH_GIT_TIMESTAMP || "",
    dirty: process.env.INARIWATCH_GIT_DIRTY === "true",
  }
}

/**
 * Extract git info at build time (runs in Node.js during the build step of
 * whichever framework is wrapping capture — Next, Nuxt, Remix, Vite, webpack,
 * etc.). Used by framework plugins to inject git context as env vars.
 */
export function extractGitInfo(): Record<string, string> {
  // Browser guard — `child_process` doesn't exist there. Bundlers only try
  // to resolve `require("child_process")` when it's a static literal, so we
  // hide it behind indirect eval (`(0, eval)("require")`).
  if (typeof window !== "undefined") return {}
  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const req = (0, eval)("require") as (m: string) => any
    const { execSync } = req("child_process")
    const run = (cmd: string): string => {
      try { return execSync(cmd, { encoding: "utf8", stdio: ["pipe", "pipe", "pipe"] }).trim() }
      catch { return "" }
    }

    const message = run("git log -1 --format=%s").slice(0, 200)
    // Scrub potential secrets from commit message
    const safeMessage = message
      .replace(/(?:sk|pk|api|key|token|secret|password)[_-]?\S{8,}/gi, "[REDACTED]")
      .replace(/:\/\/[^:]+:[^@]+@/g, "://[REDACTED]@")

    return {
      INARIWATCH_GIT_COMMIT: run("git rev-parse HEAD"),
      INARIWATCH_GIT_BRANCH: run("git rev-parse --abbrev-ref HEAD"),
      INARIWATCH_GIT_MESSAGE: safeMessage,
      INARIWATCH_GIT_TIMESTAMP: run("git log -1 --format=%cI"),
      INARIWATCH_GIT_DIRTY: run("git status --porcelain").length > 0 ? "true" : "false",
    }
  } catch {
    return {}
  }
}

File: go/git.go
// Git context — INARIWATCH_GIT_* env vars (build-time fast path) with
// subprocess `git` fallback cached once per process.
package capture

import (
	"context"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"sync"
	"time"
)

var (
	gitOnce  sync.Once
	gitCache *GitContext
)

func runGit(args ...string) string {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	out, err := exec.CommandContext(ctx, "git", args...).Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

var (
	msgSecretPattern  = regexp.MustCompile(`(?i)(?:sk|pk|api|key|token|secret|password)[_-]?\S{8,}`)
	msgConnectionPattern = regexp.MustCompile(`://[^:]+:[^@]+@`)
)

func scrubCommitMessage(msg string) string {
	out := msgSecretPattern.ReplaceAllString(msg, "[REDACTED]")
	out = msgConnectionPattern.ReplaceAllString(out, "://[REDACTED]@")
	return out
}

func fromEnv() *GitContext {
	commit := os.Getenv("INARIWATCH_GIT_COMMIT")
	if commit == "" {
		return nil
	}
	return &GitContext{
		Commit:    commit,
		Branch:    envOr("INARIWATCH_GIT_BRANCH", "unknown"),
		Message:   os.Getenv("INARIWATCH_GIT_MESSAGE"),
		Timestamp: os.Getenv("INARIWATCH_GIT_TIMESTAMP"),
		Dirty:     os.Getenv("INARIWATCH_GIT_DIRTY") == "true",
	}
}

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

func fromSubprocess() *GitContext {
	commit := runGit("rev-parse", "HEAD")
	if commit == "" {
		return nil
	}
	branch := runGit("rev-parse", "--abbrev-ref", "HEAD")
	if branch == "" {
		branch = "unknown"
	}
	msg := scrubCommitMessage(runGit("log", "-1", "--format=%s"))
	if len(msg) > 200 {
		msg = msg[:200]
	}
	dirty := runGit("status", "--porcelain") != ""
	return &GitContext{
		Commit:    commit,
		Branch:    branch,
		Message:   msg,
		Timestamp: runGit("log", "-1", "--format=%cI"),
		Dirty:     dirty,
	}
}

// GetGitContext returns build-time git metadata. Env vars take
// precedence; otherwise ``git`` is spawned once and the result cached.
func GetGitContext() *GitContext {
	gitOnce.Do(func() {
		if g := fromEnv(); g != nil {
			gitCache = g
			return
		}
		gitCache = fromSubprocess()
	})
	return gitCache
}
